pkgs/by-name/ld/ld64/meson.build

```diff
@@ -0,0 +1,248 @@
+    dispatch_queue_t queue = DISPATCH_APPLY_AUTO;
+    return 0;
+}
+'''
+if not cc.compiles(
+    dispatch_apply_auto_test,
+    args : '-Wno-unused-command-line-argument',
+    name : 'supports DISPATCH_APPLY_AUTO'
+)
+    add_project_arguments(
+        '-include', 'dispatch/private.h',
+        '-DDISPATCH_APPLY_AUTO=DISPATCH_APPLY_CURRENT_ROOT_QUEUE',
+        language: ['c', 'cpp']
+    )
+endif
+
+# The return type of `dispatch_get_global_queue` was changed in 10.14.
+# Use the older type if the SDK does not support it.
+dispatch_queue_global_test = '''
+#include <dispatch/dispatch.h>
+int main(int argc, char* argv[]) {
+    dispatch_queue_global_t queue = dispatch_get_global_queue(QOS_CLASS_USER_INITIATED, 0);
+    return 0;
+}
+'''
+if not cc.compiles(
+    dispatch_queue_global_test,
+    args : '-Wno-unused-command-line-argument',
+    name : 'supports dispatch_queue_global_t'
+)
+    add_project_arguments('-Ddispatch_queue_global_t=dispatch_queue_t', language : ['c', 'cpp'])
+endif
+
+
+# Generated files
+
+compile_stubs_h = custom_target(
+    'compile_stubs.h',
+    capture : true,
+    command : [python, '@INPUT0@', '@INPUT1@'],
+    input : ['gen_compile_stubs.py', 'compile_stubs'],
+    output : ['compile_stubs.h']
+)
+
+configure_h = custom_target(
+    'configure_h',
+    command : ['bash', '@INPUT@'],
+    env : {
+        'DERIVED_FILE_DIR' : meson.current_build_dir(),
+        'RC_ProjectSourceVersion': '@version@'
+    },
+    input : ['src/create_configure'],
+    output : ['configure.h']
+)
+
+incdirs = include_directories(
+    'compat',
+    'include',
+    'src/abstraction',
+    'src/ld',
+    'src/ld/code-sign-blobs',
+    'src/ld/parsers',
+    'src/ld/passes',
+    'src/mach_o',
+)
+
+# Dynamic libraries
+libcodedirectory = library(
+    'codedirectory',
+    dependencies : [openssl],
+    include_directories : incdirs,
+    install : true,
+    sources : [
+        'compat/corecrypto/ccdigest.c',
+        'compat/corecrypto/ccsha1.c',
+        'compat/corecrypto/ccsha2.c',
+        'compat/libcodedirectory.c',
+        'src/ld/libcodedirectory.c'
+    ],
+    soversion : 1,
+)
+install_headers(
+    'src/ld/cs_blobs.h',
+    'src/ld/libcodedirectory.h',
+)
+
+
+# Static libraries
+libprunetrie = static_library(
+    'prunetrie',
+    include_directories : incdirs,
+    install : true,
+    override_options : {'b_lto': false},
+    sources : [
+        'src/mach_o/Error.cpp',
+        'src/mach_o/ExportsTrie.cpp',
+        'src/other/PruneTrie.cpp'
+    ]
+)
+install_headers(
+    'src/other/prune_trie.h',
+    subdir : 'mach-o'
+)
+
+
+# Binaries
+ld64 = executable(
+    '@targetPrefix@ld',
+    c_args : ['-DPRIVATE'],
+    cpp_args : ['-DPRIVATE'],
+    dependencies : [libtapi, openssl, xar],
+    include_directories : incdirs,
+    install : true,
+    link_args : [
+        '-Wl,-exported_symbol,__mh_execute_header',
+        '-Wl,-stack_size,0x10000000',
+        '-Wl,-client_name,ld'
+    ],
```

Comment:
where are these args coming from? if i look at `project.pbxproj`  for the release of ld `F933D91D09291AC90083EAC8` or ld64 `F933D92509291AC90083EAC8` they don't seem to match up. perhaps i'm looking in the wrong place.

Reply:
They’re located under `OTHER_LDFLAGS`.

https://github.com/apple-oss-distributions/ld64/blob/47f477cb721755419018f7530038b272e9d0cdea/ld64.xcodeproj/project.pbxproj#L1212-L1221

<img width="1037" alt="xcode flags" src="https://github.com/NixOS/nixpkgs/assets/7413633/ab3acf0f-1a29-4e0f-938f-2ad024ae5374">

Reply:
ok. well, they are not matching up. the stack size is off and that listing doesn't seem to list the warnings or the preprocessor definitions, eg `BUILDING_LD=1` and i guess `NDEBUG` is getting handled by the meson flag?

Reply:
The stack size in the link matches, but it’s apparently that’s the debug build. I’ll update it to use the stack size from the release build.

`NDEBUG` is handled by Meson. `BUILDING_LD` isn’t referenced outside of the Xcode project, so it doesn’t make sense to define it. Are there any other preprocessor definitions I missed?

Reply:
if this is the build we are trying to mimic then maybe?

https://github.com/apple-oss-distributions/ld64/blob/47f477cb721755419018f7530038b272e9d0cdea/ld64.xcodeproj/project.pbxproj#L1235-L1315

that is F933D91D09291AC90083EAC8 in the project.pbxproj. there are lines like:

```
GCC_SYMBOLS_PRIVATE_EXTERN = NO;
CLANG_ENABLE_OBJC_WEAK = YES;
GCC_INLINES_ARE_PRIVATE_EXTERN = NO;
GCC_DYNAMIC_NO_PIC = YES;
```
which i assume map to compiler flags which i don't see in the meson file.

there is also a ld64 in that file which uses different flags `F933D92509291AC90083EAC8`.

perhaps something in the meson file saying which build the flags are coming from so it is easier to map to the provided project.pbxproj.

Reply:
I’ll add comments explaining the source of the defines as well as the usage of `-DPRIVATE`, which is needed to use private libdispatch APIs. (While I normally do not like using private APIs, they became public in later releases.)

Regarding the listed configuration, which I believe is Xcode build settings not preprocessor definitions:

* [`GCC_SYMBOLS_PRIVATE_EXTERN`][1] exports symbols by default. Setting this to `YES` is the equivalent of `-fvisibility=hidden`. `NO` is the default for clang.
* `CLANG_ENABLE_OBJC_WEAK` is the equivalent of `-fobjc-weak`. It enables ARC-style weak references in Objective-C. ld64 does not contain any Objective-C code.
* [`GCC_INLINES_ARE_PRIVATE_EXTERN`][2] apparently makes non-inline copies of inlined functions be marked private extern. This is set to `NO`, which AFAIK is the default in clang.
* [`GCC_DYNAMIC_NO_PIC`][3] builds dynamic libraries as non-PIC. Darwin has been PIE by default since 10.7, and aarch64-darwin _only_ supports PIE.

  While the Xcode project suggests that PIE is disabled, the version of ld-classic shipped by Apple appears to be PIE on both platforms.
  
  ```shell
  $ nix run nixpkgs#file /Library/Developer/CommandLineTools/usr/bin/ld-classic
  /Library/Developer/CommandLineTools/usr/bin/ld-classic: Mach-O universal binary with 2 architectures:   [x86_64:\012- Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|BINDS_TO_WEAK|PIE>] [\012- arm64:\012- Mach-O 64-bit arm64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|BINDS_TO_WEAK|PIE>]
  ```

[1]: https://developer.apple.com/library/archive/documentation/DeveloperTools/Reference/XcodeBuildSettingRef/1-Build_Setting_Reference/build_setting_ref.html#//apple_ref/doc/uid/TP40003931-CH3-SW95
[2]: https://xs-labs.com/en/blog/2011/02/04/xcode-build-settings/
[3]: https://developer.apple.com/library/archive/documentation/DeveloperTools/Reference/XcodeBuildSettingRef/1-Build_Setting_Reference/build_setting_ref.html#https://developer.apple.com/library/archive/documentation/DeveloperTools/Reference/XcodeBuildSettingRef/1-Build_Setting_Reference/build_setting_ref.html#//apple_ref/doc/uid/TP40003931-CH3-SW95

Reply:
I fixed the stack size and added a comment with a link explaining the source of the defines. I also added a comment about the need to define `PRIVATE` for libdispatch.

Reply:
Actually, since I already check for old APIs via a configure check, I moved the `-DPRIVATE` there, so it will only be used with older SDKs.